### PR TITLE
VAN-2543 Added lambda function app deployment template

### PR DESCRIPTION
**Pull request by pranaySinghDev**
Uses serverless framework to deploy functions
Uses API gateway based lambda deployment template
Uses a container artifact based deployment template

### CODE REVIEW

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+      type: string
+      default: "1024"
+    timeout:
+      title: timeout
+      description: timeout in seconds
+      type: integer
+      default: 10
+    project:
+      title: Project Name
+      description: This is a vanguard project name
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:37]: We should stop using "Vanguard" in the interfaces - change to Code Pipes

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      type: string
+      default: latest
+    applicationName:
+      title: Application name
+      description: This is vanguard application name
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:42]: ditto

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+      type: string
+      default: latest
+    environment:
+      title: Environment Name
+      description: This is vanguard environment name
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:47]: ditto

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+      type: string
+      default: dev
+    architecture:
+      title: CPU architecture
+      description: This could be arm64 or amd_x64
+      type: string
+      default: arm64
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:54]: seems like the default should be amd, no?

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    bucket:
+      title: bucket to store state of IAC
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:56]: Why is this "IAC"? is that what serverless is generating?

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 6]
```diff
@@ -0,0 +1,152 @@
+provisioner: aws
+name: lambda-gateway
+version: 1
+revision: 1
+displayName: "AWS Lambda deployment"
+description: "deploy lambda functions with codepipes and serverless"
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:6]: change to "Code Pipes"

Comment (dshelley66) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:6]: Is the fact that we are using "serverless" material to the consumer?

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+                    path: {{ func.path }}
+                    method: {{ func.verb }}
+          {% endfor %}  
+          plugins:
+            - serverless-deployment-bucket  
+          " >> serverless.yml
+      {% if job_type == 'deploy' %}
+        - serverless deploy
+      {% else %}
+        - sls remove
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:151]: why is it "serverless deploy" but "sls remove"?

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+author: "CloudCover"
+meta:
+  inputArtifactType:
+    - ContainerImage
+inputs:
+  type: object
+  properties:
+    region:
+      title: Region name
+      description: region name for cluster
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:18]: what "cluster"?

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+      type: string
+    app_env:
+      title: Application Environment
+      description: Application Related Envs
+      type: object
+      default: {}
+    memory:
+      title: Memory
+      description: Memory to be consumed by app runner
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:27]: "app runner"?

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+inputs:
+  type: object
+  properties:
+    region:
+      title: Region name
+      description: region name for cluster
+      type: string
+    app_env:
+      title: Application Environment
+      description: Application Related Envs
```
Comment (dshelley66) [inline pipeline-modules/deployment-service/aws/lambda-gateway.yaml:22]: I'm assuming this is for environment variables? If so, the description should be more descriptive:
"Application related environment variables"

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml (hunk 4 above)
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:54]: yes

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml (hunk 6 above)
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:6]: not exactly unless we state what this template is capable of doing

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml (hunk 7 above)
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:151]: both works, its an alias. i change it to be one

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml (hunk 8 above)
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:18]: by mistake, removed

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml (hunk 9 above)
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:27]: by mistake, removed

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml (hunk 5 above)
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml:56]: Serverless is generating cloud formation templates and state files that will be stored in a bucket, that way we can deploy and remove the same resource based on serverless. yaml spec without adding this information to our databases.

File: pipeline-modules/deployment-service/aws/lambda-gateway.yaml [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+      default: arm64
+    bucket:
+      title: bucket to store state of IAC
+      description: S3 bucket to hold serverless state file
+      type: string
+    funcs:
+      title: function config
+      description: Function configuration
+      type: array
+      default: []
```
[the remark this answers is not in the document]
Comment (pranaySinghDev) [reply pipeline-modules/deployment-service/aws/lambda-gateway.yaml]: added validation

**Pull request merged**